client.go

```diff
@@ -0,0 +1,400 @@
+	for _, lg := range lgs {
+		pblgs = append(pblgs, &pb.GetRecordsRequest_LogEntry{
+			LogID:  &pb.ProtoPeerID{ID: lg.ID},
+			Offset: &pb.ProtoCid{Cid: offsets[lg.ID]},
+			Limit:  int32(limit),
+		})
+	}
+
+	req := &pb.GetRecordsRequest{
+		Header: &pb.GetRecordsRequest_Header{
+			From: &pb.ProtoPeerID{ID: s.threads.host.ID()},
+		},
+		ThreadID: &pb.ProtoThreadID{ID: id},
+		Logs:     pblgs,
+	}
+
+	lg, err := s.threads.store.LogInfo(id, lid)
+	if err != nil {
+		return nil, err
+	}
+	if lg.PubKey == nil {
+		return nil, fmt.Errorf("log not found")
+	}
+
+	// Pull from each address
+	recs := make(map[peer.ID]*records)
+	wg := sync.WaitGroup{}
+	for _, addr := range lg.Addrs {
+		wg.Add(1)
+		go func(addr ma.Multiaddr) {
+			defer wg.Done()
+			p, err := addr.ValueForProtocol(ma.P_P2P)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			pid, err := peer.IDB58Decode(p)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+			if pid.String() == s.threads.host.ID().String() {
+				return
+			}
+
+			log.Debugf("getting records from %s...", p)
+
+			cctx, cancel := context.WithTimeout(ctx, reqTimeout)
+			defer cancel()
+			conn, err := s.dial(cctx, pid, grpc.WithInsecure())
+			if err != nil {
+				log.Errorf("dial %s failed: %s", p, err)
+				return
+			}
+			client := pb.NewThreadsClient(conn)
+			reply, err := client.GetRecords(cctx, req)
+			if err != nil {
+				log.Error(err)
+				return
+			}
+
+			for _, l := range reply.Logs {
+				log.Debugf("received %d records in log %s from %s", len(l.Records), l.LogID.ID.String(), p)
+
+				lg, err := s.threads.store.LogInfo(id, l.LogID.ID)
+				if err != nil {
+					log.Error(err)
+					return
+				}
+				if lg.PubKey == nil {
+					if l.Log != nil {
+						lg = logFromProto(l.Log)
+						if err = s.threads.store.AddLog(id, lg); err != nil {
+							log.Error(err)
+							return
+						}
+					} else {
+						continue
+					}
+				}
+
+				if recs[lg.ID] == nil {
+					recs[lg.ID] = newRecords()
+				}
+
+				for _, r := range l.Records {
+					rec, err := cbor.RecordFromProto(r, lg.FollowKey)
+					if err != nil {
+						log.Error(err)
+						return
+					}
+					recs[lg.ID].Store(rec.Cid(), rec)
+				}
+			}
+		}(addr)
+	}
+
+	wg.Wait()
+
+	res := make(map[peer.ID][]thread.Record)
+	for lid, rs := range recs {
+		res[lid] = rs.List()
+	}
+	return res, nil
+}
+
+// pushRecord to log addresses and thread topic.
+func (s *service) pushRecord(
+	ctx context.Context,
+	rec thread.Record,
+	id thread.ID,
+	lid peer.ID,
+	settings *tserv.AddSettings,
+) error {
+	// Collect known writers
+	var addrs []ma.Multiaddr
+	info, err := s.threads.store.ThreadInfo(settings.ThreadID)
+	if err != nil {
+		return err
+	}
+	for _, l := range info.Logs {
+		laddrs, err := s.threads.store.Addrs(settings.ThreadID, l)
+		if err != nil {
+			return err
+		}
+		addrs = append(addrs, laddrs...)
+	}
+
+	// Add additional addresses
+	addrs = append(addrs, settings.Addrs...)
```

Comment:
Sugg: Move to `addrs` definition with `len(settings.Addrs)` as min capacity

Reply:
nice, adding